### PR TITLE
[networking] Add response extensions for impersonate info

### PR DESCRIPTION
**Pull request by bashonly**
this can be useful during extractor error handling, [e.g. in the crunchyroll extractor](https://github.com/bashonly/yt-dlp/commit/e1ff1e1dff52ecda258d2a8fe61f29d786ad0881), where we want to expect the 403 error and a offer a cloudflare bypass hint if impersonation was not used, or else raise w/ a bug report message if impersonation was used.

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check all of the following options that apply:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)

### What is the purpose of your *pull request*?
- [x] Core bug fix/improvement

</details>

### CODE REVIEW

File: yt_dlp/networking/_curlcffi.py [hunk 1]
```diff
@@ -178,6 +177,8 @@ def _send(self, request: Request):
         session.curl.setopt(CurlOpt.LOW_SPEED_LIMIT, 1)  # 1 byte per second
         session.curl.setopt(CurlOpt.LOW_SPEED_TIME, math.ceil(timeout))
 
+        impersonate_target = self._get_request_target(request)
```
Comment (coletdjnz) [inline yt_dlp/networking/_curlcffi.py:180]: It should use the mapped target imo

**Review by coletdjnz — CHANGES_REQUESTED** [review]

**Comment by pukkandan** [comment]
Pls explain motivation for https://github.com/yt-dlp/yt-dlp/pull/9756/commits/3b48105b8c2a6034f97d00d3bf876d90773218f4. I liked the first impl much better

**Comment by bashonly** [comment]
@pukkandan 

hypothetically, the value of `impersonate` can be `None`:
https://github.com/yt-dlp/yt-dlp/blob/52f5be1f1e0dc45bb397ab950f564721976a39bf/yt_dlp/networking/impersonate.py#L61-L80

so if a response is an instance of `ImpersonateResponse`, then that is completely meaningless on its own. In order to determine if impersonation was used for a request, the extractor would also have to do check truthiness of the `impersonate` attribute, with the code looking like this:

```py
from ..networking.impersonate import ImpersonateResponse

try:
    ...
except ExtractorError as e:
    if isinstance(e.cause, HTTPError) and e.cause.status == 403:
        if isinstance(e.cause.response, ImpersonateResponse) and e.cause.response.impersonate is not None:
            target = str(e.cause.response.impersonate)
            # handling for impersonation
        # handling for no impersonation
```

vs. with new impl:

```py
try:
    ...
except ExtractorError as e:
    if isinstance(e.cause, HTTPError) and e.cause.status == 403:
        if target := e.cause.response.extras.get('impersonate'):
            # handling for impersonation
        # handling for no impersonation
```

File: yt_dlp/networking/common.py [hunk 2]
```diff
@@ -517,6 +517,7 @@ def __init__(
             self.reason = reason or HTTPStatus(status).phrase
         except ValueError:
             self.reason = None
+        self.extensions = {}
```
Comment (coletdjnz) [inline yt_dlp/networking/common.py:520]: I would also expose this as a parameter for Response, same as Request

+ add in docstring

**Review by Grub4K — APPROVED** [review]

**Pull request merged**